src/lib/json-rpc-router.ts

```diff
@@ -0,0 +1,37 @@
+import type { Request } from 'express';
+import type { Readable } from 'readable-stream';
+import type { JsonRpcRequest, JsonRpcResponse } from './json-rpc.js';
+
+type HttpRequest = {
+  contentType: 'application/octet-stream' | 'multipart/form-data';
+  request: Request;
+};
+
+export type RequestContext = {
+  transport?: 'http' | 'ws';
+} & (HttpRequest | { [K in keyof HttpRequest]?: never; });
```

Comment:
does this mean "a `HttpRequest` or an object that contains anything except for the keys in `HttpRequest`"?  if so, what are we really gaining here by doing this?

Reply:
it makes it so that if `request` is provided then `contentType` must also be provided and vice versa

Reply:
ah sorry i should've been clearer.  i was referring to the `{ [K in keyof HttpRequest]?: never; }` part

wouldnt `HttpRequest` by itself be enough to say "both must be provided"?

ooohhh wait is this doing something like "either you must provide both or neither"?  if so, would you mind adding a comment above this just so that if someone not super familiar with TS comes along (like me) they aren't so confused?